### PR TITLE
Rewrite assignments

### PR DESCRIPTION
**Pull request by yogan**

### CODE REVIEW

**Review by RussellSnyder — APPROVED** [review]
good idea to handle null states. One of the big advantages of Typescript

**Comment by yogan** [comment]
> good idea to handle null states. One of the big advantages of Typescript

Yeah, big *strict null checks* fanboy here.

We don't have that TS option activated in the assignment (and shouldn't),
but I always try to code like it is anyway.

File: assignment-3.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+```typescript
+test.skip('Climate app shows the current temperature…', …)
+```
+
+## 3.2 Header
+
+Let's get started by writing some tests for the `Header` component that we
+have implemented in the first assignment. You have probably put that into a
+file called `Header.tsx`. Create a new file `Header.test.tsx` for the tests.
```
Comment (seriousManual) [inline assignment-3.md:32]: maybe add a remark that says "by convention the test files have the same name as the component file"
if this is not mentioned in th slides.

File: assignment-3.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+Try to `render` the component with different text values that you pass in as a
+prop. Use `getByRole` and `toHaveTextContent` to find the rendered header in the
+DOM and check its text content.
+
+*Optional:* If you're feeling fancy, you can use
+[Jest's `test.each`](https://jestjs.io/docs/en/api#testeachtablename-fn-timeout)
+to create a parametrized test.
```
Comment (seriousManual) [inline assignment-3.md]: I'd drop that paragraph, probably only confuses the participants

File: assignment-3.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+## 3.3 Current Temperature and Humidity
+
+Back to `Climate.test.tsx`. Remove that `skip` you added in 3.1 and get the test
+working again!
+
+When it works, add a similar test for the current humidity value.
+
+### Hints
+
+You have introduced a child component in assignment 3.2 and are using that in
```
Comment (seriousManual) [inline assignment-3.md:62]: wasn't the child component added in assignment 2 somewhere?
3.2 is about testing the header.

File: assignment-3.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+## 3.5 Reset Button
+
+Test that the reset button is doing what it should. Also make sure that after
+the button is pressed the minimum and maximum are being updated again when new
+values are emitted by the sensor.
+
+Try to avoid using a `data-testid` this time. Query for the button with
+`findByRole` and a `name` option. The *Testing Playground* extension can help
+you again. If both of your buttons have the same text, change that! It makes
```
Comment (seriousManual) [inline assignment-3.md]: the test implies that there should be two reset buttons, while the assignment 2 does not specifically mention this.
(it even explicitly states "add a reset button")

File: assignment-3.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+your app more accessible, e.g. to screen readers.
+
+Again, it is sufficient to do this for either temperature or humidity.
+
+### Hints
+
+Use the [`user-event`](https://github.com/testing-library/user-event) lib to
```
Comment (seriousManual) [inline assignment-3.md:96]: good hint!

File: assignment-4.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
+# Assignment 4: Application State with Redux
+
+Up to now you have structured your application in a way that holds the state of
+the individual components locally. You learned how you can use React to create a
```
Comment (seriousManual) [inline assignment-4.md]:
```suggestion
the individual components locally. You learned how you can use Redux to create a
```

File: assignment-2.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+## 2.2 Add Minimum and Maximum
+
+In addition to current temperature and humidity, the minimum and maximum of
+those values over time shall be shown.
+
+Adapt your new component so that it keeps track of the minimum and maximum in
+its state. Make sure that after each new emitted value from the sensor, the
+current, minimum, and maximum values are updated.
+
+Notice how each instance of the component has its own independant state.
```
Comment (seriousManual) [inline assignment-2.md]:
```suggestion
Notice how each instance of the component has its own independent state.
```

File: assignment-2.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+This is a refactoring task, so the functionality should not change! You can even
+use the existing test (run with `npm test`) to be sure.
+
+## 2.2 Add Minimum and Maximum
+
+In addition to current temperature and humidity, the minimum and maximum of
+those values over time shall be shown.
+
+Adapt your new component so that it keeps track of the minimum and maximum in
```
Comment (seriousManual) [inline assignment-2.md]: The assignment "drives" people to keep track of the state inside the child components, I don't think we should push them in that direction that explicitly. 
I think it's better to let them come up with their own solution. Maybe we have more than one approach, which we can talk about later one in the assignment wrap-up.

(also I think keeping the state in the parent is a better solution)

File: assignment-3.md (hunk 3 above)
Comment (seriousManual) [reply assignment-3.md:62]: it's 2.1 I think

**Review by seriousManual — CHANGES_REQUESTED** [review]
some smol remarks

File: assignment-3.md (hunk 3 above)
Comment (yogan) [reply assignment-3.md:62]: Already solved in a later commit.

File: assignment-3.md (hunk 1 above)
Comment (yogan) [reply assignment-3.md:32]: I'll put a bullet point for the on the slides. Here I think it should be implicitly clear,
as it says "you have `Header.tsx`, create `Header.test.tsx`", I think everybody should
get from that to also use this naming for further test files.

Comment (yogan) [reply assignment-3.md:32]: Tracking with issue 65 in our slides board, closing this thread.

File: assignment-3.md (hunk 4 above)
Comment (yogan) [reply assignment-3.md]: Good catch!

File: assignment-3.md (hunk 2 above)
Comment (yogan) [reply assignment-3.md]: ACK

Comment (yogan) [reply assignment-3.md]: Done.

File: assignment-3.md (hunk 4 above)
Comment (yogan) [reply assignment-3.md]: I've just removed it.

File: assignment-2.md (hunk 8 above)
Comment (yogan) [reply assignment-2.md]: You are right. I've shortened it to just the task to be done, without given any impl. details:
https://github.com/methodpark/react-workshop/compare/5c9e58f347d833dcd2c74f479fab3dd59f45e9a4..d7ae133c42e27af4319fbad4ccd1ffde3a3cfb74

**Review by seriousManual — APPROVED** [review]

**Comment by seriousManual** [comment]
let's do it!

**Pull request merged**